vxyowsup/whatsapp.py

```diff
@@ -109,29 +113,36 @@ def _kill():
 
     def send_to_stack(self, text, to_address):
         def send():
-            self.echo_layer.send_to_human(text, to_address)
-            print('in stack_client', text, to_address)
+            self.whatsapp_interface.send_to_human(text, to_address)
         self.stack.execDetached(send)
 
 
-class EchoLayer(YowInterfaceLayer):
+class WhatsAppInterface(YowInterfaceLayer):
+
+    def __init__(self, transport):
+        super(WhatsAppInterface, self).__init__()
+        self.transport = transport
 
     def send_to_human(self, text, to_address):
         message = TextMessageProtocolEntity(text, to=to_address)
         self.toLower(message)
-        # print('in echo_layer', text, to_address)
 
     @ProtocolEntityCallback("message")
     def onMessage(self, messageProtocolEntity):
+        from_address = messageProtocolEntity.getFrom()
+        body = messageProtocolEntity.getBody()
+        to_address = messageProtocolEntity.getTo()
+
         receipt = OutgoingReceiptProtocolEntity(
             messageProtocolEntity.getId(),
-            messageProtocolEntity.getFrom(), 'read',
+            from_address, 'read',
             messageProtocolEntity.getParticipant())
 
         self.toLower(receipt)
 
-        self.send_to_human(text=messageProtocolEntity.getBody(),
-                           to_address=messageProtocolEntity.getFrom())
+        reactor.callFromThread(self.transport.publish_message,
+                               from_addr=from_address, content=body, to_addr=to_address,
```

Comment:
So here we would strip the `@s.whatsapp.net` from `from_address` before sending it out. It might be polite to also set the `to_addr_type` to `TransportUserMessage.AT_MSISDN`.